### PR TITLE
Add PR refinement loop with code review integration

### LINKED ISSUE #33: Add PR refinement loop with code review integration
## Summary

Extend the Ralph Loop to include automated code review and PR refinement, enabling fully unattended workflow from implementation through merge.

## Problem

Currently, after LXA completes milestone tasks and creates a draft PR, humans must:
1. Wait for CI and fix issues
2. Run code reviews manually
3. Fix review feedback
4. Mark PR ready for review
5. Handle automated review comments
6. Squash & merge

This interrupts the autonomous workflow and requires human attention at multiple points.

## Proposed Solution

Add a **refinement loop** that runs after tasks complete:

1. Wait for CI → fix failures
2. Run code review (using codereview-roasted principles)
3. Parse verdict (🟢 Good taste / 🟡 Acceptable / 🔴 Needs rework)
4. If not clean: implement fixes, commit, push, repeat
5. When clean: mark PR ready, optionally auto-merge

## CLI Interface

```
lxa implement [DESIGN_DOC] [OPTIONS]

Arguments:
  DESIGN_DOC          Path to design document (default: .pr/design.md)

Options:
  --loop              Run continuously until all milestones complete
  --refine            Run code review refinement loop after tasks complete
  --auto-merge        Squash & merge when refinement passes
  --allow-merge TYPE  Stop refinement on: good_taste | acceptable (default: acceptable)
  --min-iterations N  Minimum review iterations before accepting (default: 1)
  --max-iterations N  Maximum review iterations (default: 5)
```

## Examples

```bash
# Current behavior (stop at draft PR)
lxa implement --loop

# Refine until clean, stop at ready-for-review
lxa implement --loop --refine

# Full automation through merge
lxa implement --loop --refine --auto-merge

# Strict: require good taste
lxa implement --loop --refine --auto-merge --allow-merge good_taste

# Pragmatic: accept "acceptable" after 2 passes
lxa implement --loop --refine --auto-merge --allow-merge acceptable --min-iterations 2
```

## Stop Conditions

| Verdict | Behavior |
|---------|----------|
| 🟢 Good taste | Always stop (code is clean) |
| 🟡 Acceptable | Stop if `allow_merge=acceptable` AND `iteration >= min_iterations` |
| 🔴 Needs rework | Always continue (must fix critical issues) |
| Any | Stop if `iteration >= max_iterations` (safety limit) |

## Implementation Approach

Minimal changes using existing tools:
1. Add CLI arguments to `src/__main__.py`
2. Pass config through `RalphLoopRunner`
3. Add orchestrator skill with refinement loop instructions
4. Track iteration state in `.pr/refinement-state.json`

No new tools required - orchestrator delegates code review to sub-agent and uses `gh` CLI for PR operations.

### PR DESCRIPTION
**Pull request by jpshackelford**
## Summary

Extends LXA with automated PR refinement through a **two-phase code review loop**:

1. **Phase 1: Self-Review** - Agent reviews its own code, fixes issues, gets PR ready for external review
2. **Phase 2: Review Response** - Agent reads external review comments, addresses them, and marks threads resolved

Closes #33

### Phase 1: Self-Review (Pre-Review)

When the PR is a draft or hasn't been reviewed yet:

```
┌─────────────────────────────────────────┐
│         PHASE 1: SELF-REVIEW            │
└─────────────────────────────────────────┘
                    │
                    ▼
         Wait for CI to pass
                    │
                    ▼
         Delegate: Code Review
         (codereview-roasted style)
                    │
                    ▼
         Returns: 🟢/🟡/🔴 + issues
                    │
            ┌───────┴───────┐
            │               │
         🟢/🟡            🔴
            │               │
            ▼               ▼
      Mark PR Ready    Fix issues
                       Commit & push
                       Loop back
```

**Goal**: Get the code clean enough for human review.

### Phase 2: Review Response (Post-Review)

When the PR has unresolved review comments:

```
┌─────────────────────────────────────────┐
│       PHASE 2: REVIEW RESPONSE          │
└─────────────────────────────────────────┘
                    │
                    ▼
         Read unresolved review threads
                    │
                    ▼
         For each comment:
           1. Understand the ask
           2. Make the fix
           3. Commit with reference
           4. Reply to thread
           5. Mark thread resolved
                    │
                    ▼
         Push & wait for CI
                    │
                    ▼
         Loop until no unresolved threads
                    │
                    ▼
         Auto-merge (if configured)
```

**Goal**: Address reviewer feedback systematically until PR is approved.

## Usage

```bash
# Auto-detect phase based on PR state
lxa refine https://github.com/owner/repo/pull/42

# Explicitly run self-review phase
lxa refine https://github.com/owner/repo/pull/42 --phase self-review

# Explicitly run review response phase
lxa refine https://github.com/owner/repo/pull/42 --phase respond

# Full automation through merge
lxa refine https://github.com/owner/repo/pull/42 --auto-merge
```

## CLI Options

```
lxa refine PR_URL [OPTIONS]

Arguments:
  PR_URL                GitHub PR URL (e.g., https://github.com/owner/repo/pull/42)

Options:
  --phase TYPE          Phase to run: auto | self-review | respond (default: auto)
  --auto-merge          Squash & merge when refinement passes
  --allow-merge TYPE    Quality bar: good_taste | acceptable (default: acceptable)
  --min-iterations N    Minimum review iterations before accepting (default: 1)
  --max-iterations N    Maximum refinement iterations (default: 5)
```

## Design Principles

| Aspect | Approach |
|--------|----------|
| **State Management** | Python code controls the loop, not LLM prompts |
| **Thread Resolution** | Uses GitHub GraphQL API to reply and resolve |
| **CI Integration** | Polls/waits for checks before proceeding |
| **Testability** | Each phase is a discrete, testable unit |
| **Visibility** | Progress tracked via resolved threads |

## Implementation

| File | Purpose |
|------|---------|
| `src/__main__.py` | CLI with `--phase` argument |
| `src/ralph/refine.py` | `RefineRunner` with two-phase loop |
| `src/ralph/github_review.py` | GitHub API helpers for review threads |
| `src/agents/orchestrator.py` | Orchestrator skill for integrated mode |

## Testing

- Unit tests for URL parsing
- Unit tests for review thread parsing
- Integration tests for each phase
- All existing tests pass

### CODE REVIEW

File: src/__main__.py [hunk 1]
```diff
@@ -267,30 +270,148 @@ def run_reconcile(design_doc: Path, workspace: Path, *, dry_run: bool = False) -
     return 0
 
 
-def run_ralph_loop(design_doc: Path, workspace: Path, *, max_iterations: int = 20) -> int:
+def parse_pr_url(pr_url: str) -> tuple[str, int]:
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/__main__.py]: Fixed in 4277acd - Moved parse_pr_url() to src/utils/github.py to improve separation of concerns. CLI module now focuses on wiring, not business logic.

File: src/ralph/refine.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+Refine PR #{self.pr_number} using the refinement loop.
+
+Configuration:
+- allow_merge: {self.refinement_config.allow_merge}
+- min_iterations: {self.refinement_config.min_iterations}
+- max_iterations: {self.refinement_config.max_iterations}
+- auto_merge: {self.refinement_config.auto_merge}
+
+Start by checking out the PR branch and running the refinement loop.
+Output REFINEMENT_COMPLETE when done.
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/ralph/refine.py]: Fixed in feb6512 - Replaced fragile magic string parsing with robust case-insensitive pattern matching. Now supports multiple completion indicators (phase_complete, finished, done, etc.) to handle LLM output variations.

File: src/ralph/refine.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+Your task is to refine PR #{pr_number} in repository {repo_slug} until it meets the quality bar.
+
+REFINEMENT LOOP:
+1. Check out the PR branch: `gh pr checkout {pr_number} --repo {repo_slug}`
+2. Wait for CI to complete: `gh pr checks {pr_number} --repo {repo_slug} --watch`
+3. If CI fails: fix the issues, commit, push, restart loop
+4. Read iteration count from .pr/refinement-state.json (create if missing)
+5. Increment iteration and save back to state file
+6. Perform code review focusing on:
+   - Data structures and simplicity (not style)
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/ralph/refine.py]: Fixed in a1998ec - Consolidated duplicated refinement logic into shared refinement_config.py. Now there is ONE source of truth for workflows, code review principles, and commit guidelines. Changes to refinement logic only need to be made in one place.

File: src/__main__.py [hunk 4]
```diff
@@ -373,6 +494,34 @@ def main(argv: list[str] | None = None) -> int:
         default=20,
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/__main__.py]: Fixed in ccb6e1f - Added comprehensive test suite with 39 new tests:

- **tests/utils/test_github.py**: 7 tests for parse_pr_url function (URL validation, edge cases)
- **tests/test_cli_refine.py**: 6 tests for CLI argument parsing and imports
- **tests/ralph/test_refine_runner.py**: 13 tests for RefineRunner core logic (phase detection, completion detection, conversation output)
- **tests/ralph/test_state.py**: 13 tests for state management classes (StateManager, RefinementState)

All tests pass and cover the new CLI functionality, URL parsing, and core refinement logic.

File: tests/ralph/test_refine_runner.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        """Test conversation output extraction with agent messages."""
+        from openhands.sdk.event import MessageEvent
+
+        runner = RefineRunner(
+            self.mock_llm, self.workspace, self.pr_number, self.repo_slug, self.refinement_config
+        )
+
+        # Create mock events
+        mock_event1 = Mock(spec=MessageEvent)
+        mock_event1.source = "agent"
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply tests/ralph/test_refine_runner.py]: Fixed in 9389584f2af55327659df36ac0edee7b7e20cdee

Replaced `Mock(spec=MessageEvent)` with real `MessageEvent` and `Message` objects from the OpenHands SDK. The test now uses proper constructors and validates actual behavior. This will catch real bugs if the event structure or content handling changes.

**Comment by jpshackelford** [comment]
@openhands One of our prompts seems to suggest that we should be responding to review threads on the PR before we push. That is incorrect. We should update review threads after pushing the relevant changes.
